doc/images/ipsec_fsm.gv

```diff
@@ -0,0 +1,32 @@
+digraph ipsec_state_machine {
```

Comment:
I do not quite understand the SA state machine picture, maybe because I think it tries to include different kind of states (global IPsec subsystem state, IPsec SA state, IPsec operation state) in the same picture.

The operation related states (Processing, Op_Complete) are not SA states but states of individual packet operations. There can be multiple packet operations ongoing and in different states at the same time for the same SA.

I can identify only the following four SA states:

1) Nonexistent: The SA does not exist
- odp_ipsec_sa_create() changes the SA state to SA_Ready.

2) SA_Ready: The SA can be used for processing packets
- odp_ipsec_sa_disable() changes the SA state to Disable_Pending

3) Disable_Pending: The SA is processing in-flight packets but cannot anymore be passed as a parameter to any new IPsec packet operation.
- The ODP implementation removes the SA from inbound SA lookup (if the SA is an inbound SA that offloads SA lookup) and completes processing the in-flight packets for the SA and enqueues the operation completion IPsec packet events. After that the ODP implementation enqueues a disable complete SA status event and then the SA state changes to SA_Disabled.

4) SA_Disabled: The SA has been disabled.
- No packets are processed by the SA and no new IPsec packet or status events for the SA appear.
- odp_ipsec_sa_destroy() destroys the SA fully and changes the SA state to Nonexistent.

Then one could define SA_expired and SA_not_expired sub-states for some of the states (SA_Ready and Disable_Pending), but the lifetime related states could also be seen as a separate FSM, mostly independent from the main SA FSM.

Reply:
You're right, the diagram is trying to represent the IPsec subsystem as well as individual SAs. I can separate it out into multiple FSMs if that would be clearer. The point is to show which ODP APIs are valid at various stages of processing. So, for example, once an SA is expired you cannot initiate new IPsec operations with it, etc.